pkg/reconciler/sample/samplesource.go

```diff
@@ -131,9 +109,162 @@ func (r *Reconciler) ReconcileKind(ctx context.Context, src *v1alpha1.SampleSour
+			return nil
+		}
+
+		// Check if there is an old ip, so we can cleanup that conn
+		r.srcPodsIPsLock.Lock()
+		if oldIP, ok := r.srcPodsIPs[string(src.UID)]; ok && oldIP != podIp {
+			r.controlConnections.RemoveConnection(ctx, string(src.UID))
+		}
+		// Update with new one
+		r.srcPodsIPs[string(src.UID)] = podIp
+		r.srcPodsIPsLock.Unlock()
+
+		ctrl, err = r.controlConnections.CreateNewControlInterface(ctx, string(src.UID), podIp)
+		if err != nil {
+			return fmt.Errorf("cannot connect to the pod: %w", err)
+		}
+
+		// We need to start the message listener for this control interface
+		deploymentName := ra.Name
+		srcName := src.Name
+		srcNamespace := src.Namespace
+		go func() {
+			for ev := range ctrl.InboundMessages() {
+				r.controlMessageHandler(ctx, ev.Event(), deploymentName, srcName, srcNamespace)
+				ev.Ack()
+			}
+		}()
+	}
+
+	logging.FromContext(ctx).Infof("we have a control connection")
+
+	// --- If last configuration update to that deployment doesn't contain the interval, set it
+	actualInterval, _ := time.ParseDuration(src.Spec.Interval) // No need to check the error here, the webhook already did it
+	if old, ok := r.getLastSentInterval(ra); !ok || old != actualInterval {
+		event := cloudevents.NewEvent()
+		event.SetID(uuid.New().String())
+		event.SetType("updateinterval.samplesource.control.knative.dev")
+		event.SetSource("samplesource-controller")
+		event.SetExtension("newinterval", src.Spec.Interval)
+
+		err := ctrl.SendAndWaitForAck(event)
+		if err != nil {
+			return fmt.Errorf("cannot send the event to the pod: %w", err)
+		}
+
+		r.lastIntervalUpdateSentLock.Lock()
+		r.lastIntervalUpdateSent[ra.Name] = actualInterval
+		r.lastIntervalUpdateSentLock.Unlock()
+	}
+
+	logging.FromContext(ctx).Infof("we have sent the interval update")
+
+	src.Status.MarkConfigurationNotPropagated()
+
+	// --- If last status update to that deployment doesn't contain the correct interval,
+	// it means we still didn't received that status update message
+	// TODO here we're using interval for simplicity,
+	//  but we may need to find another way to uniquely identify a configuration (hash of the in memory object?)
+	if ackedInterval, ok := r.getLastUpdate(ra); !ok || ackedInterval != actualInterval {
+		return nil
+	}
+
+	logging.FromContext(ctx).Infof("we have received the status update")
+
+	src.Status.MarkConfigurationPropagated()
+
+	return nil
+}
+
+func (r *Reconciler) FinalizeKind(ctx context.Context, src *v1alpha1.SampleSource) pkgreconciler.Event {
+	// Check if there is an old ip, so we can cleanup that conn
+	r.srcPodsIPsLock.Lock()
+	if _, ok := r.srcPodsIPs[string(src.UID)]; ok {
+		r.controlConnections.RemoveConnection(ctx, string(src.UID))
+	}
+	// Update with new one
+	delete(r.srcPodsIPs, string(src.UID))
+	r.srcPodsIPsLock.Unlock()
+
 	return nil
 }
 
+func (r *Reconciler) controlMessageHandler(ctx context.Context, event cloudevents.Event, deploymentName string, srcName string, srcNamespace string) {
```

Comment:
I'd make this its own type and not implement it on the reconciler. From a structural PoV, this has nothing to do with the reconciler lifecycle per-se but runs asynchronous. It's prolly a good idea to model it that way.